opt_adapt/opt.py

```diff
@@ -212,7 +206,7 @@ def _gradient_descent(it, forward_run, m, params, u, u_, dJ_):
         u_ = params.transfer_fn(u_, u.function_space())
         dJ_diff = fd.assemble(ufl.inner(dJ_ - dJ, dJ_ - dJ) * ufl.dx)
         lr = abs(fd.assemble(ufl.inner(u_ - u, dJ_ - dJ) * ufl.dx) / dJ_diff)
-
+    lr = max(lr, params.lr_min)
```

Comment:
Nice. I think we should do this in the other methods, too. Do you agree?

Reply:
Yep, I modify the code in line search, and the other 4 methods all need line search to choose lr.

Reply:
Oh of course.